extendr-api/tests/rarray_tests.rs

```diff
@@ -0,0 +1,19 @@
+//! This test module contains tests for [RArray], [RMatrix], etc.
+//!
+use extendr_api::prelude::*;
+use extendr_engine::with_r;
+
+#[test]
+fn flat_index_test() {
+    with_r(|| {
+        let example = R!("matrix(as.numeric(1:9), byrow=TRUE, ncol = 3, nrow = 3)").unwrap();
+        let test_array = RArray::<f64, ()>::try_from(&example).unwrap();
```

Comment:
In this case, wouldn't it make more sense to make an `RArray<f64,[usize;1]>` and have an index trait for that constraint? 

Alternatively, could we have a `flat_index()` method instead? Or is the `[]` notation super important? I guess from my perspective if you have to cast to `RArray<f64, ()>` what is the difference to just using `as_real_slice()`

Reply:
It was a big mistake to add this dimension thing to `RArray`/`RMatrix`. It is inspired by `ndarray`. One thing that `ndarray` does in the case of run-time specified indices, is it has https://docs.rs/ndarray/latest/ndarray/fn.IxDyn.html, which is more closer to what we have with R's matrix stuff.

I think we really only know that `D = ()`, as it is a runtime thing, and can be changed, unless you solemnly swear, that you won't change it. But that's not a real rust API. Pinky promises aren't a thing here.

I think we should rectify the situation as much, and introduce the use `()` in the API more and more. For instance, it feels like `Dataframe<T>` can only be used when you have a row-type `T`. But that's not true. There is a lot of advantages from using `Dataframe<()>`, mainly the checks are simpler. 

I think we pre-emptively cluttered the design of these things.. If something should be special cased, it should be this `D` thing, and the column-wise flat index should be the standard.

Reply:
Yes, I'm definitely thinking those things. 

I think you're right, this is a little too loose patch. A refactor would be more appropriate. 
Let me think about it and propose another PR.

Reply:
Whatever is in there already is strange. For instance, I couldn't find an impl for accepting `array()` from R. So it is incomplete.